src/include/miopen/invoke_params.hpp

```diff
@@ -0,0 +1,168 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/errors.hpp>
+
+#include <memory>
+#include <typeinfo>
+#include <type_traits>
+#include <utility>
+
+namespace miopen {
+
+enum class InvokeType
+{
+    Run,
+    Evaluate,
+    AutoTune,
+};
+
+struct InvokeParams
+{
+    InvokeType type = InvokeType::Run;
+};
+
+struct AnyInvokeParams
+{
+    public:
+    AnyInvokeParams() = default;
+
+    template <class Actual>
+    AnyInvokeParams(const Actual& value) : impl(std::make_unique<Implementation<Actual>>(value))
+    {
+    }
+
+    template <class Actual,
+              class = std::enable_if_t<!std::is_same<Actual, AnyInvokeParams>::value, void>>
+    AnyInvokeParams(Actual&& value)
+        : impl(std::make_unique<Implementation<Actual>>(std::forward(value)))
+    {
+    }
+
+    AnyInvokeParams(const AnyInvokeParams& other) : impl(other.impl ? other.impl->Copy() : nullptr)
+    {
+    }
+
+    AnyInvokeParams(AnyInvokeParams&& other) noexcept : impl(std::move(other.impl)) {}
+
+    AnyInvokeParams& operator=(const AnyInvokeParams& other)
+    {
+        if(&other != this)
+            impl = other.impl ? other.impl->Copy() : nullptr;
+        return *this;
+    }
+
+    AnyInvokeParams& operator=(AnyInvokeParams&& other) noexcept
+    {
+        if(&other != this)
+            impl = std::move(other.impl);
+        return *this;
+    }
+
+    void SetInvokeType(InvokeType type)
+    {
+        if(!impl)
+            MIOPEN_THROW("Attempt to use empty AnyInvokeParams.");
+        impl->SetInvokeType(type);
+    }
+
+    InvokeType GetInvokeType() const
+    {
+        if(!impl)
+            MIOPEN_THROW("Attempt to use empty AnyInvokeParams.");
+        return impl->GetInvokeType();
+    }
+
+    template <class Actual>
+    const Actual& CastTo() const
+    {
+        if(!impl)
+            MIOPEN_THROW("Attempt to use empty AnyInvokeParams.");
+        if(!impl->CanCastTo(typeid(Actual)))
+            MIOPEN_THROW("Attempt to cast AnyInvokeParams to invalid type.");
+        return *reinterpret_cast<Actual*>(impl->GetRawPtr());
+    }
+
+    template <class Actual>
+    Actual& CastTo()
```

Comment:
This should be a static or free function so you dont have to create multiple overloads for const and non-const.

Reply:
To be honest, I don't understand how being static/free would help with multiple overloads. I could make it copying like boost::any_cast, which would reduce it to a single const overload, or make it templated (regarding receiving AnyInvokeParams) with auto, which would be a strange solution to eliminate a single five-lined overload. Could you please clarify?

Reply:
I was suggesting writing something like this:

```cpp
    template<class Actual, class T>
    static Actual& AnyCast(T&& x)
    {
        if(!x.impl)
            MIOPEN_THROW("Attempt to use empty AnyInvokeParams.");
        if(!x.impl->CanCastTo(typeid(Actual)))
            MIOPEN_THROW("Attempt to cast AnyInvokeParams to invalid type.");
        return *reinterpret_cast<Actual*>(x.impl->GetRawPtr());
    }
```

But this will require the `const` to match. It might be better to have two seperate overloads and remove const on the `const` overload.